### ISSUE TITLE
the code of mm_models.py

### ISSUE THREAD

**Issue by KeepMoving97** [PaulCCCCCCH/Multimodal-Categorization-of-Crisis-Events-in-Social-Media#3]
hello, thanks for your code. I have some questions about the code of mm_models.py. Could you answer me the questions? Thank you very much.
1. In the line 92, line 122, and line 127, why you used the Dropout()?  The f_i and e_i should be a vector or a matrix, but if you used the Dropout(), they will no be the vector or matrix.

**Comment by KeepMoving97** [comment]
I only run the mm_models.py because i just need the attention. But when i debug in the line 127, it raise a error "alueError: expected 2D or 3D input (got 1D input)". The demension of e_i is (768, ), it could not be fed in BatchNormalize(). And it is also different of the f_i, because the dimension of f_i is (N, 1000).

**Comment by PaulCCCCCCH** [comment]
Hi, 
Thanks for asking. I'm not sure if I understand correctly where your confusions are, but I'll try to give some clarifications:

- Dropout does not change the shape of the input. Line 61 initializes the dropout layer, and then you can simply feed input with any shape into it, and it will zero-out random entries of the input according to the given probability at initialization. 
- If you only run `mm_models.py`, you need to make sure that the input is properly batched, i.e., the argument `x` given to the `forward` method of `DenseNetBertMMModel` should be a tuple of image batch (N images) and text batch (N text) (see line 118). Given your description, the most likely cause is that the text you gave to the model was not batched, i.e., you used shape (T, dim_text_repr), but the code expected (N, T, dim_text_repr). Same applies to image. I would recommend looking at how these data is pre-processed and loaded in `crisismmd_dataset.py` and `preprocessing.py`.

If you need further clarification, feel free to ask again!

Thanks.

**Comment by KeepMoving97** [comment]
Thank you very much for your answers!

**Comment by KeepMoving97** [comment]
Dear autor, i would like to ask if my data shape is (N, D) (the N denotes the sample number, and the D denotes the dimension of data) how can i fed in the DenseNetBertMMModel()? I will change densenet and bert to the network that suits my data.
Thank you very much for your answers again!

**Comment by PaulCCCCCCH** [comment]
I'm afraid you can't, because it requires input to be a (text, image) tuple. And this makes sense because this is an end-to-end model, you give in image and text, it gives a prediction back. If you give it a vector (likely to be some middle product), it won't work.

**Comment by KeepMoving97** [comment]
Maybe I didn't express it clearly. The data I input is also a tuple corresponding to the image and text, but I am not doing this direction. I am going to feed the data of two different modalities of my research direction into the model. . But my data is only 2D(batch, D), not 3D(batch, T, D) like images and text.

**Comment by PaulCCCCCCH** [comment]
Oh I think I made a mistake. The input text I used for text came from [Huggingface BERT tokenizer](https://huggingface.co/docs/transformers/model_doc/bert#transformers.BertTokenizer). You can try getting your input text with the following 
```
from transformers import BertTokenizer
import torch

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")

text = tokenizer("Hello, my dog is cute")
```
Here you get one input text. You can add this logic in your `__getitem__` function of your data loader, it will convert text into a batch for you. Be aware that `text` here is a dictionary of values (see the document for  BertTokenizer), so I transform the batch further in `trainer.py` line 63 - 65.

**Comment by KeepMoving97** [comment]
thanks



| |
***@***.***
|
|
***@***.***
|




---- 回复的原邮件 ----
| 发件人 | ***@***.***> |
| 日期 | 2022年05月07日 04:20 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.******@***.***> |
| 主题 | Re: [PaulCCCCCCH/Multimodal-Categorization-of-Crisis-Events-in-Social-Media] the code of mm_models.py (Issue #3) |

Oh I think I made a mistake. The input text I used for text came from Huggingface BERT tokenizer. You can try getting your input text with the following

from transformers import BertTokenizer
import torch

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")

text = tokenizer("Hello, my dog is cute")


Here you get one input text. You can add this logic in your __getitem__ function of your data loader, it will convert text into a batch for you. Be aware that text here is a dictionary of values (see the document for BertTokenizer), so I transform the batch further in trainer.py line 63 - 65.

**Comment by KeepMoving97** [comment]
Thanks!
